File: 2021/23.cs
using Aoc.Library.Puzzles;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Aoc2021;
public class Day23 : Puzzle
{
    public static int SIZE = 2;

    public override int Day => 23;

    public override string SolvePartA(string[] input)
    {
        Console.WriteLine();
        var sw = new Stopwatch();
        SIZE = 2;
        var init = new State(new Bucket[] {
            new Bucket(SIZE, 'B', 'A'),
            new Bucket(SIZE, 'C', 'D'),
            new Bucket(SIZE, 'D', 'B'),
            new Bucket(SIZE, 'A', 'C')
        });

        //return PlayManually(init);

        sw.Start();
        List<long> solutions = Solve(init, useAssumption: true);
        sw.Stop();
        Console.WriteLine(sw.ElapsedMilliseconds);
        return solutions.First().ToString();
    }

    public override string SolvePartB(string[] input)
    {
        Console.WriteLine();
        var sw = new Stopwatch();
        SIZE = 4;
        var init = new State(new Bucket[] {
            new Bucket(SIZE, 'B', 'D', 'D', 'A'),
            new Bucket(SIZE, 'C', 'B', 'C', 'D'),
            new Bucket(SIZE, 'D', 'A', 'B', 'B'),
            new Bucket(SIZE, 'A', 'C', 'A', 'C')
        });
        sw.Start();
        List<long> solutions = Solve(init);
        sw.Stop();
        Console.WriteLine(sw.ElapsedMilliseconds);
        return solutions.First().ToString();
    }

    private List<long> Solve(State init, bool useAssumption = false)
    {
        const bool isDfs = true;
        var q = new Stack<State>();
        q.Push(init);

        var solutions = new List<long>();

        var seen = new HashSet<string>();
        var i = 0;
        int previousStep = 0;
        while (q.Count > 0)
        {
            i++;
            var state = q.Pop();
            if (i % 100000 == 0) // state.steps > previousStep)
            {
                previousStep = state.steps;
                Console.WriteLine($"{i}, {solutions.Count}, {state.steps}, {q.Count}\r");
            }
            //Console.WriteLine(state);
            seen.Add(state.GetHashKey(isDfs));
            foreach (var move in state.GenMoves(useAssumption))
            {
                var newState = state.Clone();
                move(newState);
                if (IsSorted(newState))
                {
                    solutions.Add(newState.energyConsumed);
                    continue;
                }
                if (!seen.Contains(newState.GetHashKey(isDfs))
                    // max steps for optimal solution cannot exceed it
                    // The rule for move generation should enforce this, but for some reason they don't in DFS.
                    && state.steps <= SIZE * 4 * 2) 
                {
                    q.Push(newState);
                }
            }
        }

        solutions.Sort();
        return solutions;
    }

    private string PlayManually(State state)
    {
        var history = new List<State>();
        while(true)
        {
            Console.WriteLine();
            Console.WriteLine("- - - - - - - - - - - - - - - - - - - -");
            Console.WriteLine(state.ToString());
            if (IsSorted(state))
            {
                Console.WriteLine("Congrats!");
                break;
            }
            Console.WriteLine("Next Step: [q]uit, [p]revious, in <slot> <bucket>, out <bucket> <slot>");
            var answer = Console.ReadLine();
            try
            {
                if (answer == null || answer.Length == 0 || answer[0] == 'q') break;
                else if (answer[0] == 'p')
                {
                    state = history[history.Count - 1];
                    history.RemoveAt(history.Count - 1);
                }
                else if (answer[0] == 'i')
                {
                    history.Add(state);
                    state = state.Clone();
                    var parts = answer.Split(' ');
                    state.MoveIn(int.Parse(parts[2]), int.Parse(parts[1]));
                }
                else if (answer[0] == 'o')
                {
                    history.Add(state);
                    state = state.Clone();
                    var parts = answer.Split(' ');
                    state.MoveOut(int.Parse(parts[1]), int.Parse(parts[2]));
                }
                else
                {
                    Console.WriteLine("Unable to parse answer: " + answer);
                }
            }
            catch (Exception e) {
                Console.WriteLine("Error: " + e.Message);
            };
        }
        return "";
    }



    private class Bucket : IEnumerable<char>
    {
        private Stack<char> pods;
        public readonly int size;

        public Bucket(int size, params char[] pods)
        {
            this.pods = new Stack<char>(pods);
            this.size = size;
        }

        public Bucket Clone() => new(size, pods.Reverse().ToArray());

        public (char, int) Take()
        {
            var distance = this.size - pods.Count;
            var c = pods.Pop();
            return (c, distance);
        }

        public int Put(char c)
        {
            if (this.IsFull()) { throw new InvalidOperationException("Bucket is full"); }

            pods.Push(c);
            var distance = this.size - pods.Count; // 0 for top spot, 1 for bottom
            return distance;
        }

        public bool IsDone(char expected)
        {
            return this.pods.Count == this.size
                && this.pods.All(x => x == expected);
        }

        public bool IsEmpty() => this.pods.Count == 0;
        public bool IsFull() => this.pods.Count == this.size;

        public char Peek()
        {
            return this.pods.Peek();
        }

        public override int GetHashCode()
        {
            return this.pods.GetHashCode();
        }

        public IEnumerator<char> GetEnumerator() => this.pods.GetEnumerator();
        IEnumerator IEnumerable.GetEnumerator() => this.pods.GetEnumerator();
    }

    private class State
    {
        public Bucket[] buckets = new Bucket[4];
        public long energyConsumed = 0;
        public char[] spaces = new char[7];
        public int steps = 0;

        public State(Bucket[] buckets)
        {
            this.buckets = buckets;
        }

        public IEnumerable<Action<State>> GenMoves(bool useAssumption)
        {
            var expected = new char[] {'A', 'B', 'C', 'D'};

            bool couldMoveIn = false;
            for(var slot = 0; slot <= 6; slot++)
            {
                if (this.spaces[slot] != 0)
                {
                    for(var i = 0; i < this.buckets.Length; i++)
                    {
                        var goal = expected[i];
                        var bucket = this.buckets[i];

                        if (!bucket.IsFull()
                            && this.spaces[slot] == goal  // only move into slots you wanna be
                            && bucket.All(x => x == goal)) // only move if the slot has the right pods
                        {
                            couldMoveIn = true;
                            yield return (State state) => state.MoveIn(i, slot);
                        }
                    }
                }
            }

            // Only move stuff out, if we can't move anything in its place.
            if (couldMoveIn)
            {
                yield break;
            }

            for(var slot = 0; slot <= 6; slot++)
            {
                if (this.spaces[slot] == 0)
                {
                    for(var i = 0; i < this.buckets.Length; i++)
                    {
                        var goal = expected[i];
                        var bucket = this.buckets[i];
                        // Only take things out from buckets that are not done yet.

                        if (!bucket.IsEmpty()
                            && !bucket.IsDone(goal))
                        {
                            var pod = bucket.Peek();
                            if (!useAssumption || pod != 'D' || slot >= 3)
                                yield return (State state) => state.MoveOut(i, slot);
                        }
                    }
                }
            }

        }

        public bool IsBucketDone(int bucket)
        {
            var expected = new char[] {'A', 'B', 'C', 'D'}[bucket];
            return this.buckets[bucket].IsDone(expected);
        }

        public string GetHashKey(bool forDfs = true)
        {
            var builder = new StringBuilder();
            for(var i = 0; i < this.spaces.Length; i++)
            {
                builder.Append(i);
                builder.Append(this.spaces[i] == 0 ? '.' : this.spaces[i]);
            }
            foreach(var bucket in buckets)
            {
                builder.Append(' ');
                builder.Append(new string(bucket.Select(x => x).ToArray()));
            }
            if (forDfs)
                builder.Append(this.energyConsumed);
            return builder.ToString();
        }

        public State Clone()
        {
            var state = new State(buckets.Select(bucket => bucket.Clone()).ToArray());

            state.energyConsumed = this.energyConsumed;
            state.spaces = (char[])this.spaces.Clone();
            state.steps = this.steps;

            return state;
        }

        private int[] Range(int from, int to)
        {
            if (from > to)
            {
                (to, from) = (from, to);
            }

            return Enumerable.Range(from, to - from + 1).ToArray();
        }

        private Dictionary<char, int> price = new()
        {
            { 'A',    1 },
            { 'B',   10 },
            { 'C',  100 },
            { 'D', 1000 },
        };

        public void MoveOut(int bucket, int slot)
        {
            if (this.buckets[bucket].IsEmpty())
            {
                //Console.WriteLine($"Illegal move (bucket empty): bucket: {bucket} slot: {slot}");
                return; // nothing to move
            }

            int[] spacesToMovePast;
            // moving out to the left
            if (slot - 1 <= bucket)
            {
                spacesToMovePast = Range(slot - 1, bucket).Select(x => x + 1).ToArray();
            }
            else
            {
                spacesToMovePast = Range(bucket + 2, slot);
            }

            // if any of the spaces are occupied, then we can't move to our slot
            if (spacesToMovePast.Any(slot => this.spaces[slot] != 0))
            {
                //Console.WriteLine($"Illegal move (path blocked): slot: {spacesToMovePast.First(s => this.spaces[s] != 0)}");
                return;
            }

            int extra = (slot == 0 || slot == 6) ? -1 : 0;

            (var pod, var distance) = this.buckets[bucket].Take();
            this.energyConsumed += price[pod] * (2 * spacesToMovePast.Length + extra + distance);
            this.spaces[slot] = pod;

            this.steps++;
        }

        public void MoveIn(int bucket, int slot)
        {
            if (this.buckets[bucket].IsFull())
            {
                //Console.WriteLine($"Illegal move (bucket full): bucket: {bucket} slot: {slot}");
                return;
            }

            int[] spacesToMovePast;
            // moving in to the left
            if (slot - 1 <= bucket)
            {
                spacesToMovePast = bucket == slot - 1
                    ? new int[0]
                    : Range(slot, bucket).Select(x => x + 1).ToArray();
            }
            else
            {
                spacesToMovePast = bucket + 2 == slot 
                    ? new int[0]
                    : Range(bucket + 2, slot - 1);
            }

            // if any of the spaces are occupied, then we can't move to our slot
            if (spacesToMovePast.Any(slot => this.spaces[slot] != 0))
            {
                //Console.WriteLine($"Illegal move (path blocked): slot: {spacesToMovePast.First(s => this.spaces[s] != 0)}");
                return;
            }

            int extra = (slot == 0 || slot == 6) ? -1 : 0;

            var pod = this.spaces[slot];
            var distance = this.buckets[bucket].Put(pod);
            this.energyConsumed += price[pod] * (2 * (spacesToMovePast.Length + 1) + extra + distance);
            this.spaces[slot] = (char)0;

            this.steps++;
        }

        public override string ToString()
        {
            var text = new StringBuilder();
            var state = this.Clone();
            text.AppendLine($"Energy: {state.energyConsumed}, steps: {state.steps}");
            text.AppendLine($"01 2 3 4 56");
            for(var i = 0; i < 7; i++)
            {
                var c = state.spaces[i];
                if (c == 0) text.Append('.');
                else text.Append(c);
                if (i >= 1 && i <= 4) text.Append('.');
            }
            text.AppendLine();
            for(var i = SIZE - 1; i >= 0; i--)
            {
                text.Append("  ");
                foreach(var bucket in state.buckets)
                {
                    var lastIndex = bucket.Count() - 1;
                    if (i > lastIndex) text.Append(" ");
                    else
                    {
                        (char c, int _) = bucket.Take();
                        text.Append(c);
                    }
                    text.Append(" ");
                }
                text.AppendLine();

            }
            text.AppendLine($"  0 1 2 3");

            return text.ToString();
        }
    }


    private bool IsSorted(State state)
    {
        return state.buckets[0].IsDone('A')
            && state.buckets[1].IsDone('B')
            && state.buckets[2].IsDone('C')
            && state.buckets[3].IsDone('D');
    }

}

File: 2021/23.js
require('./aoc.js');

const SENTINEL = '.';
const expected = ['A', 'B', 'C', 'D'];
const init = {
    buckets: [
        ['B', 'A'],
        ['C', 'D'],
        ['D', 'B'],
        ['A', 'C']
    ],
    bucketSize: 2,
    energyConsumed: 0,
    corridor: [SENTINEL].repeat(7),
    steps: 0
}

// for testing purposes
const done = {
    buckets: [
        ['A', 'A'],
        ['B', 'B'],
        ['C', 'C'],
        ['D', 'D']
    ],
    bucketSize: 2,
    energyConsumed: 0,
    corridor: [SENTINEL].repeat(7),
    steps: 0
}

const init2 = {
    buckets: [
        ['B', 'D', 'D', 'A'],
        ['C', 'B', 'C', 'D'],
        ['D', 'A', 'B', 'B'],
        ['A', 'C', 'A', 'C']
    ],
    bucketSize: 4,
    energyConsumed: 0,
    corridor: [SENTINEL].repeat(7),
    steps: 0
}
/*
01  2  3  4  56
  B0 B1 B2 B3
*/

function range(a, b) {
    if (a > b) [a, b] = [b, a];
    const r = [];
    for(let i = a; i <= b; i++) r.push(i);
    return r;
}

function slotsBetween(slot, bucket) {
    if (slot - 1 <= bucket) {
        // slot + 1, because we don't want to check the given slot itself
        // bucket + 1, because that's the first slot left of the bucket
        return slot - 1 == bucket ? [] : range(slot + 1, bucket + 1);
    } else {
        // bucket + 2, because that is the first slot right of the bucket
        // slot - 1, because we don't want to check the given slot itself
        return bucket + 2 > slot - 1 ? [] : range(bucket + 2, slot - 1);
    }
}

const isPathFree = (state, slot, bucket) => slotsBetween(slot, bucket).every(i => state.corridor[i] == SENTINEL);
const isBucketEmpty = (state, bucket) => !state.buckets[bucket].length;
const isBucketFull  = (state, bucket) => state.buckets[bucket].length == state.bucketSize;
const isSlotTaken = (state, slot) => state.corridor[slot] != SENTINEL;

function moveOut(state, bucket, slot) {
    if (isBucketEmpty(state, bucket)) throw new Error('Bucket is empty');
    if (!isPathFree(state, slot, bucket)) throw new Error('Path is blocked');

    const newState = state.clone();

    const corridorDistance = 2 * (slotsBetween(slot, bucket).length + 1);
    const extra = slot == 0 || slot == 6 ? -1 : 0;
    const bucketDistance = state.bucketSize - state.buckets[bucket].length;

    const pod = newState.buckets[bucket].pop();
    newState.energyConsumed += (corridorDistance + extra + bucketDistance) * (10 ** expected.indexOf(pod));
    newState.corridor[slot] = pod; // pop from the newState, so the buckets are updated in the right state
    newState.steps++;
    return newState;
}

function moveIn(state, slot, bucket) {
    if (isBucketFull(state, bucket)) throw new Error('Bucket is full');
    if (!isPathFree(state, slot, bucket)) throw new Error('Path is blocked');

    const newState = state.clone();

    const corridorDistance = 2 * (slotsBetween(slot, bucket).length + 1);
    const extra = slot == 0 || slot == 6 ? -1 : 0;
    const bucketDistance = Math.max(0, state.bucketSize - state.buckets[bucket].length - 1);
    const pod = newState.corridor[slot];

    newState.energyConsumed += (corridorDistance + extra + bucketDistance) * (10 ** expected.indexOf(pod));
    newState.buckets[bucket].push(pod);
    newState.corridor[slot] = SENTINEL;
    newState.steps++;
    return newState;
}

function genStates(state) {
    const states = [];
    for(let slot = 0; slot < 7; slot++) {
        let pod = state.corridor[slot];
        if (pod == SENTINEL) continue;

        for(let bucket = 0; bucket < 4; bucket++) {
            let goal = expected[bucket];
            if (isBucketFull(state, bucket)) continue;      // bucket is full, can't put pods in
            if (!isPathFree(state, slot, bucket)) continue; // can't reach bucket, path is blocked
            if (!state.buckets[bucket].every(p => p == goal)) continue;  // don't move things on top of pods that still need to move out
            if(pod != goal) continue;                       // the pod does not belong in this bucket

            states.push(moveIn(state, slot, bucket));
        }
    }

    // Prefer moving pods into their buckets.
    if (states.length > 0) return states;

    for(let slot = 0; slot < 7; slot++) {
        for(let bucket = 0; bucket < 4; bucket++) {
            let goal = expected[bucket];
            if (isBucketEmpty(state, bucket)) continue;      // nothing to move out
            if (isSlotTaken(state, slot)) continue;          // can't move onto slot
            if (!isPathFree(state, slot, bucket)) continue;  // can't move, path is blocked
            if (state.buckets[bucket].every(p => p == goal)) continue;  // everything in the bucket should be

            states.push(moveOut(state, bucket, slot));
        }
    }

    return states;
}

function isSorted(state) {
    for(let i = 0; i < expected.length; i++) {
        if (state.buckets[i].length != state.bucketSize
            || state.buckets[i].some(pod => pod != expected[i]))
            {
                return false;
            }
    }
    return true;
}

function show(state) {
    const print = (i) => state.corridor[i] == SENTINEL ? '.' : state.corridor[i];
    let corridorIndices = ' ' + '01 2 3 4 56' + ' ';
    let ceiling = '┌' + '─'.repeat(11) + '┐';
    let corridor = '│' + print(0) + [1,2,3,4,5].map(print).join('.') + print(6) + '│';
    let buckets = [];
    let lastIndex = state.bucketSize - 1;
    for(let i = 0; i <= lastIndex; i++) {
        let row = [];
        for(let j = 0; j < state.buckets.length; j++) {
            let pod = state.buckets[j][lastIndex - i] || ' ';
            pod = expected[j] == pod ? ('' + pod).green() : ('' + pod).red();
            row.push(pod);
        }
        buckets.push(row);
    }
    let topRow = '└─┐' + [' '].repeat(buckets[0].length).join('┬') + '┌─┘';
    let nextRows = buckets.slice(0, buckets.length).map(b =>
                 '  │' + b.join('│') + '│').join("\n");
    let lastRow ='  └' + '─'.repeat(state.buckets.length * 2 - 1) + '┘';
    let bucketIndices = '   ' + '0 1 2 3';

    return [corridorIndices, ceiling, corridor, topRow, nextRows, lastRow, bucketIndices].join("\n");
}

function solve(initialState, maxSteps) {
    const seen = {};
    const key = state => `${state.corridor.join('')}_${state.buckets.map(b => b.join('')).join(' ')}`;

    let q = [initialState];
    const sols = [];

    let i = 0;
    while(q.length) {
        if (maxSteps && i > maxSteps) {
            return;
        }
        i++;
        const state = q.pop(); // pop = DFS, shift = BFS
        if (maxSteps) {
            console.log('─'.repeat(13));
            console.log({ sols: sols.length, energy: state.energyConsumed, steps: state.steps, q: q.length });
            console.log(show(state));
        }
        if(i % 100000 == 0) {
            console.log(key(state));
            console.log(i, sols.length, state.steps, q.length);
        }
        seen[key(state)] = state.energyConsumed;
        for(let nextState of genStates(state)) {
            if (isSorted(nextState)) {
                sols.push(nextState.energyConsumed);
                continue;
            }

            let k = key(nextState);
            if (seen[k] && seen[k] <= nextState.energyConsumed) {
                continue;
            }
            q.push(nextState);
        }
        q = q.sort((a,b) => a.energyConsumed - b.energyConsumed); // for priority queue
    }

    return sols.sort();
}

let steps = process.argv[2];
let s = solve(init, steps);
console.log(s);
